### PR TITLE
Add support for forking single branch

### LINKED ISSUE #25117: Fork using only a single branch or tag
### Feature Description

Fork is an important operation in some development processes. It will be much more convenient if the resulting repository will contain only the useful references and not all the ones from the original repository.

As it is possible with [github ]( https://github.blog/changelog/2022-07-27-you-can-now-fork-a-repo-and-copy-only-the-default-branch/), it could be achieved by selecting a branch or **a tag** for the clone operation.

### Screenshots

image

### PR DESCRIPTION
**Pull request by dsseng**
Fixes #25117

Add UI for choosing branch to fork

Change default branch on single-branch forks

image

### CODE REVIEW

File: services/repository/fork.go [hunk 1]
```diff
@@ -134,9 +139,15 @@ func ForkRepository(ctx context.Context, doer, owner *user_model.User, opts Fork
 
 		needsRollback = true
 
+		var cloneCmd *git.Command = nil
+		if opts.SingleBranch != "" {
+			cloneCmd = git.NewCommand(txCtx, "clone", "--bare", "--single-branch", "--branch")
```
Comment (dsseng) [inline services/repository/fork.go]: For some reason the cloned repos keep original size whatsoever

Comment (dsseng) [reply services/repository/fork.go]: However cloning from the Gitea server results in fork being less than original, which is correct. Server FS size of the fork is approximately equal to the size of the original

**Comment by sh7dm** [comment]
Extra idea: maybe we should support the same option for migrations? It's maybe even more useful in such case when people migrate (essentially fork a repo to their instance) for development purpose

**Comment by yp05327** [comment]
I understand the point of fork arbitrary branch, but list all branches at one time may cause performance problem.
Loading branches page by page is better, but I'm not sure whether we have this feature. 🤔

**Comment by sh7dm** [comment]
> I understand the point of fork arbitrary branch, but list all branches at one time may cause performance problem. Loading branches page by page is better, but I'm not sure whether we have this feature. thinking

Other places where branch list is used do render it in once. Listing branches and putting them into a template shouldn't hog the performance a lot

File: services/repository/fork.go (hunk 1 above)
Comment (sad75) [reply services/repository/fork.go]: Thank you for this feature.
I think that being also able to select a tag instead of a branch could be very useful.

Comment (dsseng) [reply services/repository/fork.go]: I'm unsure Git itself allows for that

Edit: the only tag-related `git clone` option is `--no-tags`

Comment (sad75) [reply services/repository/fork.go]: You just have to provide the tag name to the `--branch` option.
From [git clone](https://git-scm.com/docs/git-clone#Documentation/git-clone.txt---branchltnamegt) documentation;
**`--branch` can also take tags and detaches the HEAD at that commit in the resulting repository.**

Comment (dsseng) [reply services/repository/fork.go]: Ah, alright, thanks. I'm unsure whether tags should be shown in the same selection box by default

Comment (sad75) [reply services/repository/fork.go]: Is it possible to use the standard component for selecting a ref ?
image

**Comment by denyskon** [comment]
@sh7dm Could you please resolve conflicts? I totally missed this PR, but it looks uncomplicated to review. Unfortunately we're already in feature freeze for 1.21 though...

File: templates/repo/pulls/fork.tmpl [hunk 2]
```diff
@@ -51,6 +51,26 @@
 						</div>
 						<span class="help">{{.locale.Tr "repo.fork_visibility_helper"}}</span>
 					</div>
+					<div class="inline field">
+						<label>{{.locale.Tr "repo.fork_branch"}}</label>
```
Comment (denyskon) [inline templates/repo/pulls/fork.tmpl]:
```suggestion
						<label>{{ctx.Locale.Tr "repo.fork_branch"}}</label>
```

File: templates/repo/pulls/fork.tmpl [hunk 3]
```diff
@@ -51,6 +51,26 @@
...
+						<div class="ui selection dropdown">
+							<input type="hidden" id="fork_single_branch" name="fork_single_branch" value="" required>
+							<span class="text truncated-item-container" data-value="" title="{{.locale.Tr "repo.all_branches"}}">
```
Comment (denyskon) [inline templates/repo/pulls/fork.tmpl]:
```suggestion
							<span class="text truncated-item-container" data-value="" title="{{ctx.Locale.Tr "repo.all_branches"}}">
```

File: templates/repo/pulls/fork.tmpl [hunk 4]
```diff
@@ -51,6 +51,26 @@
...
+								<span class="truncated-item-name">{{.locale.Tr "repo.all_branches"}}</span>
```
Comment (denyskon) [inline templates/repo/pulls/fork.tmpl]:
```suggestion
								<span class="truncated-item-name">{{ctx.Locale.Tr "repo.all_branches"}}</span>
```

File: templates/repo/pulls/fork.tmpl [hunk 5]
```diff
@@ -51,6 +51,26 @@
...
+							</span>
+							{{svg "octicon-triangle-down" 14 "dropdown icon"}}
+							<div class="menu">
+								<div class="item truncated-item-container" data-value="" title="{{.locale.Tr "repo.all_branches"}}">
```
Comment (denyskon) [inline templates/repo/pulls/fork.tmpl]:
```suggestion
								<div class="item truncated-item-container" data-value="" title="{{ctx.Locale.Tr "repo.all_branches"}}">
```

File: templates/repo/pulls/fork.tmpl [hunk 6]
```diff
@@ -51,6 +51,26 @@
...
+									<span class="truncated-item-name">{{.locale.Tr "repo.all_branches"}}</span>
```
Comment (denyskon) [inline templates/repo/pulls/fork.tmpl]:
```suggestion
									<span class="truncated-item-name">{{ctx.Locale.Tr "repo.all_branches"}}</span>
```

**Review by denyskon — APPROVED** [review]
I'd like for you to use the new translation structure (see my suggestions), but in general, this PR seems to look good

File: services/repository/fork.go (hunk 1 above)
Comment (lunny) [reply services/repository/fork.go]: Maybe we can change it to

```go
cloneCmd := git.NewCommand(txCtx, "clone", "--bare")
if opts.SingleBranch != "" {
    cloneCmd = cloneCmd.AddDynamicArguments("--single-branch", "--branch", opts.SingleBranch)
}
if stdout, _, err := cloneCmd.AddDynamicArguments(oldRepoPath, repoPath).
```

**Comment by lunny** [comment]
Could you paste a screenshot?

File: services/repository/fork.go (hunk 1 above)
Comment (dsseng) [reply services/repository/fork.go]: Done. `"--single-branch", "--branch"` had to be added by `AddArguments` though

**Comment by sh7dm** [comment]
image

File: routers/web/repo/pull.go [hunk 7]
```diff
@@ -180,6 +180,21 @@ func getForkRepository(ctx *context.Context) *repo_model.Repository {
...
 	}
 
+	branches, err := git_model.FindBranchNames(ctx, git_model.FindBranchOptions{
+		RepoID: ctx.Repo.Repository.ID,
+		ListOptions: db.ListOptions{
+			ListAll: true,
+		},
+		IsDeletedBranch: util.OptionalBoolFalse,
+		// Add it as the first option
+		ExcludeBranchNames: []string{forkRepo.DefaultBranch},
```
Comment (lunny) [inline routers/web/repo/pull.go]: I think it's `ctx.Repo.Repository`'s branches not `forkRepo`'s?

File: services/repository/fork.go [hunk 8]
```diff
@@ -70,14 +71,18 @@ func ForkRepository(ctx context.Context, doer, owner *user_model.User, opts Fork
 		}
 	}
 
+	defaultBranch := opts.BaseRepo.DefaultBranch
```
Comment (lunny) [inline services/repository/fork.go:74]: If opts.SingleBranch, then it should fork with all branches but not default branch?

**Comment by delvh** [comment]
Alternative design idea:
Offer a radio group with the following possibilities:
`- all`
`- default branch`
`- custom`
Where `custom` opens a checkbox group of all available branches.
I think this improves both the functionality as well as being more usable, as the two most often used values are easy to find/discern.

**Review by lunny — APPROVED** [review]

**Comment by lunny** [comment]
> Alternative design idea: Offer a radio group with the following possibilities: `- all` `- default branch` `- custom` Where `custom` opens a checkbox group of all available branches. I think this improves both the functionality as well as being more usable, as the two most often used values are easy to find/discern.

I think it's good idea but could be another PR. At least, this PR allow we fork with a single branch.

File: services/repository/fork.go [hunk 9]
```diff
@@ -134,9 +139,13 @@ func ForkRepository(ctx context.Context, doer, owner *user_model.User, opts Fork
 
 		needsRollback = true
 
+		cloneCmd := git.NewCommand(txCtx, "clone", "--bare")
+		if opts.SingleBranch != "" {
+			cloneCmd.AddArguments("--single-branch", "--branch")
+			cloneCmd = cloneCmd.AddDynamicArguments(opts.SingleBranch)
```
Comment (wxiaoguang) [inline services/repository/fork.go]:
```suggestion
			cloneCmd.AddArguments("--single-branch", "--branch").AddDynamicArguments(opts.SingleBranch)
```

**Pull request merged**